File: data/code_pairs/cpp/973.cpp
#ifndef LINE_SEARCH_SOLVER_H
#define LINE_SEARCH_SOLVER_H

#include <memory>

class line_search {
public:
    line_search() = default;
    virtual ~line_search() = default;
};

class line_search_solver {
private:
    std::unique_ptr<line_search> m_lineSearch;
    bool m_useLineSearch = false;

public:
    line_search_solver() = default;

    void get_line_search(std::unique_ptr<line_search>& ls) const {
        if (m_lineSearch) {
            ls = std::make_unique<line_search>(*m_lineSearch);
        }
    }

    void set_line_search(const line_search& ls) {
        m_lineSearch = std::make_unique<line_search>(ls);
    }

    void set_default() {
        line_search ls;
        set_line_search(ls);
    }

    bool is_line_search_defined() const {
        return bool(m_lineSearch);
    }

    bool get_use_search() const {
        return m_useLineSearch;
    }

    void set_use_search(bool x) {
        m_useLineSearch = x;
    }
};

#endif // LINE_SEARCH_SOLVER_H

File: data/code_pairs/fortran/973.f90
module nonlin_solve_line_search
    implicit none

    type :: line_search
        ! Dummy type for example.
    end type line_search

    type :: line_search_solver
        class(line_search), allocatable :: m_lineSearch
        logical :: m_useLineSearch = .false.
    contains
        procedure :: set_line_search
        procedure :: get_line_search
        procedure :: set_default
        procedure :: is_line_search_defined
        procedure :: get_use_search
        procedure :: set_use_search
    end type line_search_solver

contains

    subroutine set_line_search(this, ls)
        class(line_search_solver), intent(inout) :: this
        class(line_search), intent(in) :: ls
        if (allocated(this%m_lineSearch)) deallocate(this%m_lineSearch)
        allocate(this%m_lineSearch, source = ls)
    end subroutine set_line_search

    subroutine get_line_search(this, ls)
        class(line_search_solver), intent(in) :: this
        class(line_search), allocatable, intent(out) :: ls
        if (allocated(this%m_lineSearch)) then
            allocate(ls, source = this%m_lineSearch)
        endif
    end subroutine get_line_search

    subroutine set_default(this)
        class(line_search_solver), intent(inout) :: this
        type(line_search) :: ls
        call this%set_line_search(ls)
    end subroutine set_default

    function is_line_search_defined(this) result(res)
        class(line_search_solver), intent(in) :: this
        logical :: res
        res = allocated(this%m_lineSearch)
    end function is_line_search_defined

    function get_use_search(this) result(res)
        class(line_search_solver), intent(in) :: this
        logical :: res
        res = this%m_useLineSearch
    end function get_use_search

    subroutine set_use_search(this, val)
        class(line_search_solver), intent(inout) :: this
        logical, intent(in) :: val
        this%m_useLineSearch = val
    end subroutine set_use_search

end module nonlin_solve_line_search